oauthenticator/globus.py

```diff
@@ -174,44 +185,67 @@ def globus_portal_client(self):
+        # Repackage the Auth token into a dict that looks like the other tokens
+        auth_token_dict = {attr_name: token_json.get(attr_name) for attr_name in token_attrs}
+        # Make sure only the essentials make it into tokens. Other items, such as 'state' are
+        # not needed after authentication and can be discarded.
+        other_tokens = [{attr_name: token_dict.get(attr_name) for attr_name in token_attrs}
+                        for token_dict in token_json['other_tokens']]
+        tokens = other_tokens + [auth_token_dict]
+        # historically, tokens have been organized by resource server for convenience.
+        # If multiple scopes are requested from the same resource server, they will be
+        # combined into a single token from Globus Auth.
+        by_resource_server = {
+                    token_dict['resource_server']: token_dict
+                    for token_dict in tokens
+                    if token_dict['resource_server'] not in self.exclude_tokens
+                }
         return {
             'name': username,
             'auth_state': {
                 'client_id': self.client_id,
-                'tokens': {
-                    tok: v
-                    for tok, v in tokens.by_resource_server.items()
-                    if tok not in self.exclude_tokens
-                },
+                'tokens': by_resource_server,
             },
         }
 
-    def revoke_service_tokens(self, services):
+    def get_default_headers(self):
+        return {"Accept": "application/json", "User-Agent": "JupyterHub"}
+
+    def get_client_credential_headers(self):
+        headers = self.get_default_headers()
+        b64key = base64.b64encode(
+            bytes("{}:{}".format(self.client_id, self.client_secret), "utf8")
+        )
+        headers["Authorization"] = "Basic {}".format(b64key.decode("utf8"))
+        return headers
+
+    async def revoke_service_tokens(self, services):
         """Revoke live Globus access and refresh tokens. Revoking inert or
         non-existent tokens does nothing. Services are defined by dicts
         returned by tokens.by_resource_server, for example:
         services = { 'transfer.api.globus.org': {'access_token': 'token'}, ...
             <Additional services>...
         }
         """
-        client = self.globus_portal_client()
-        for service_data in services.values():
-            client.oauth2_revoke_token(service_data['access_token'])
-            client.oauth2_revoke_token(service_data['refresh_token'])
-
-    def get_callback_url(self, handler=None):
-        """
-        Getting the configured callback url
-        """
-        if self.oauth_callback_url is None:
-            raise HTTPError(
-                500,
-                'No callback url provided. '
-                'Please configure by adding '
-                'c.GlobusOAuthenticator.oauth_callback_url '
-                'to the config',
-            )
```

Comment:
`get_callback_url` was reimplemented by the superclass and is much better than when this was originally written, so it's actually better if it's removed entirely. The new one can correctly guess the callback url in most cases.